[IMP] account_edi_ubl_cii: support AttachedDocument imports for UBL

andg-odoo opened the pull request:
AttachedDocuments are a wrapper filetype in the UBL 2.0/2.1 specification that allows for multiple documents to be bundled together. Colombia uses it as a way to return back values from the EDI but it is in the standard and can therefore be used by any UBL 2.0/2.1 compliant system.

To be able to parse the documents inside correctly, we must obtain the original record which is stored in the outermost Attachment node either under the EmbeddedDocumentBinaryObject element or the ExternalReference/Description node. Once we find it we send it to the normal decoder process.

Specification: https://docs.oasis-open.org/ubl/os-UBL-2.1/mod/summary/reports/UBL-AttachedDocument-2.1.html

task-4299222

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/208953.png)](https://mergebot.odoo.com/odoo/odoo/pull/208953)

antoine162 commented on addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:
```diff
@@ -701,6 +703,13 @@ def _get_document_type_code_vals(self, invoice, invoice_data):
     # IMPORT
     # -------------------------------------------------------------------------
 
+    def _import_invoice_ubl_cii(self, invoice, file_data, new=False):
+        """ Allow AttachedDocument files to have their original record properly parsed. """
+        if file_data['xml_tree'].tag.split('}')[-1] == 'AttachedDocument':
+            file_data['xml_tree'] = self._ubl_parse_attached_document(file_data['xml_tree'])
+
+        return super()._import_invoice_ubl_cii(invoice, file_data, new=new)
```
I would put this logic in the `_get_edi_decoder` method in `account_move.py`, and check there whether the document is an AttachedDocument, and if it is, extract the embedded attachment *before* we call `_get_ubl_cii_builder_from_xml_tree`. That way, you can make sure that `_get_ubl_cii_builder_from_xml_tree` looks at the *inner document* to identify its type.

antoine162 commented on addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:
```diff
@@ -701,6 +703,13 @@ def _get_document_type_code_vals(self, invoice, invoice_data):
 
+    def _import_invoice_ubl_cii(self, invoice, file_data, new=False):
+        """ Allow AttachedDocument files to have their original record properly parsed. """
+        if file_data['xml_tree'].tag.split('}')[-1] == 'AttachedDocument':
```
```suggestion
        if etree.QName(file_data['xml_tree']).localname == 'AttachedDocument':
```
should normally work according to the LXML docs

https://lxml.de/3.7/tutorial.html#:~:text=tag%20%3D%20etree.QName(root)

antoine162 commented on addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:
```diff
@@ -851,6 +860,39 @@ def _correct_invoice_tax_amount(self, tree, invoice):
+
+        attachment_binary_data = attachment_node.find('./{*}EmbeddedDocumentBinaryObject')
+        if attachment_binary_data is not None \
+                and attachment_binary_data.attrib.get('encodingCode', 'base64') == 'base64' \
+                and attachment_binary_data.attrib.get('mimeCode') in ('application/xml', 'text/xml'):
```
I don't know whether these conditions are necessary. `encodingCode` for example is optional.
https://docs.oasis-open.org/ubl/os-UBL-2.1/mod/summary/reports/UBL-AttachedDocument-2.1.html#UDT-BinaryObject.Type:~:text=type%20is%20text.-,encodingCode,optional,-xsd%3AnormalizedString

Maybe we can just try to decode + parse the contents, and if that doesn't work, fall back on the description?

antoine162 reviewed (commented):
Just a quick look.
It would be nice to add tests for importing both the EmbeddedDocumentBinaryObject and the Description cases. You would need to add those tests in `l10n_account_edi_ubl_cii_tests`. See the `_assert_imported_invoice_from_file` helper e.g. https://github.com/odoo/odoo/blob/6ccdbeaf6115c22d783fc62f77e51c0fa3e0b6dc/addons/l10n_account_edi_ubl_cii_tests/tests/test_xml_cii_fr.py#L435

andg-odoo commented:
@antoine162 I've added tests for the two cases you mentioned to this PR. I wasn't sure if l10n_co should be added as a dependency to the test module so since this is a standard document that could be used like this I added it to the AU tests since it was one of the first files using UBL.

I considered adding it to a generic UBL test file but I wasn't sure how to decide what country and taxes the invoice should be in if I did that.

Let me know if there are any other changes!

antoine162 reviewed (commented):
Thanks so much for the tests and your explanations :smile: 

It would be better if those tests were separated in a separate test file which just handles testing the import of AttachedDocument. That helps make it clear that they aren't testing something Australia-specific. You could also mention in the test docstring that this has nothing to do with Australia. 

Besides that, using the Australian localization in the tests is perfectly fine as you did. Thanks so much! :smile:

andg-odoo commented:
@antoine162 Understood, test case moved to it's own class and explained in the docstrings why we used Australia. Let me know if there are any other changes necessary!

andg-odoo commented:
@antoine162 let me know if anything else needs to be changed about the test cases or implementation here :smile:

smetl commented:
@antoine162 I let you having a second look on this PR :) Let's end up this next week

antoine162 reviewed (approved):
To me it's good, sorry for not looking at it earlier!
@robodoo r+
